doc/contrib/consistency.rst

```diff
@@ -0,0 +1,62 @@
+
+******************
+Default Parameters
+******************
+
+Like many other machine learning libraries, XGBoost all of them can either be inferred from the data or have default values. Bindings should not make copies of these default values and let the XGBoost core to decide. When the parameter key is not passed into the C++ core, XGBoost will pick the default accordingly. These default are not necessarily optimal, but they are there for consistency. If there's a new choice of default parameter, we can change it inside the core and it will be automatically propagated to all bindings. Given the same set of parameters and data, various bindings should stride to produce the same model. One exception is the `num_boost_rounds`, which exists only in high level bindings and has various alias like ``n_estimators``. It's default value is close to arbitrary at the moment, we haven't been able to get a good default yet.
+
+*******
+Logging
+*******
+
+XGBoost has a default logger builtin that can be a wrapper over binding-specific logging facility. For instance, the Python binding registers a callback to use Python :py:mod:`warnings` and :py:func:`print` function to output logging. We want to keep logging native to the larger communities instead of using the ``std::cerr`` from C++.
+
+***********************************
+Minimum Amount of Data Manipulation
+***********************************
+
+XGBoost is mostly a machine learning library providing boosting algorithm implementation. Some other implementations might perform some sort of data manipulation implicitly like deciding the coding of the data, transforming the data according to some heuristic before training. We prefer to keep these operations based on necessities instead of convenience to keep the scope of the project well defined. Whenever possible, we should leave these features to 3-party libraries and consider how a user can compose their pipeline. For instance, XGBoost itself should not perform ordinal encoding for categorical data, users will pick an encoder that fits their use cases (like out-of-core implementation, distributed implementation, known mapping, etc). If some transformations are decided to be part of the algorithm, we can have it inside the core instead of the language binding. Examples would be target-encoding or sketching the response variables. If we were to support them, we can have it inside the core implementation as part of the ML algorithm. This aligns with the same principals of default parameter, various bindings should provide similar (if not the same) results given same set of parameters and data.
```

Comment:
```suggestion
XGBoost is mostly a machine learning library providing boosting algorithm implementation. Some other implementations might perform some sort of data manipulation implicitly like deciding the coding of the data, and transforming the data according to some heuristic before training. We prefer to keep these operations based on necessities instead of convenience to keep the scope of the project well-defined. Whenever possible, we should leave these features to 3-party libraries and consider how a user can compose their pipeline. For instance, XGBoost itself should not perform ordinal encoding for categorical data, users will pick an encoder that fits their use cases (like out-of-core implementation, distributed implementation, known mapping, etc). If some transformations are decided to be part of the algorithm, we can have it inside the core instead of the language binding. Examples would be target-encoding or sketching the response variables. If we were to support them, we could have it inside the core implementation as part of the ML algorithm. This aligns with the same principles of default parameters, various bindings should provide similar (if not the same) results given the same set of parameters and data.
```